### PR TITLE
Add `send_message_to_l2` Cairo Cheatcode

### PR DESCRIPTION
**Pull request by kasperski95**
Closes https://github.com/software-mansion/protostar/issues/1337

### CODE REVIEW

File: protostar/starknet/cheaters/contracts.py [hunk 1]
```diff
@@ -311,3 +308,49 @@ async def invoke(
...
+        payload: Optional[CairoOrPythonData] = None,
+    ) -> CairoData:
+        if payload is None:
+            return []
+        if isinstance(payload, collections.Mapping):
+            python_payload = {"from_address": int(l1_address), **payload}
+            return await self._transform_calldata_to_cairo_data_by_addr(
+                function_name=str(selector),
+                calldata=python_payload,
+                contract_address=contract_address,
```
Comment (Arcticae) [inline protostar/starknet/cheaters/contracts.py:354]: Perhaps we should consider transferring this functionality to another cheater, along with call, invoke - a transaction-based cheater. I'd do this in another PR though, after #1368

File: protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            payload: Optional[CairoOrPythonData] = None,
+        ) -> None:
+            asyncio.run(
+                self.cheaters.contracts.send_message_to_l2(
+                    from_l1_address=Address.from_user_input(from_address),
+                    to_l2_address=Address.from_user_input(to_address),
+                    selector=Selector(function_name),
+                    payload=payload,
+                )
+            )
```
Comment (Arcticae) [inline protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py:28]: Don't we want to catch and re-throw wrapped exceptions here?

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: protostar/starknet/cheaters/contracts.py [hunk 3]
```diff
@@ -311,3 +308,49 @@ async def invoke(
...
+        to_l2_address: Address,
+        payload: Optional[CairoOrPythonData] = None,
+    ) -> None:
+        cairo_calldata = await self._transform_payload_from_l1_to_cairo_data_by_addr(
+            l1_address=from_l1_address,
+            contract_address=to_l2_address,
+            payload=payload,
+            selector=selector,
+        )
+        entry_point = ExecuteEntryPoint.create_for_testing(
```
Comment (Arcticae) [inline protostar/starknet/cheaters/contracts.py:325]: This will not work if the contract we're calling is `warp`ed or `roll`ed or `mock`ed. We should add a test for that, and use cheatable entrypoint here.

Comment (kasperski95) [reply protostar/starknet/cheaters/contracts.py]: Changed to `CheatableExectureEntryPoint`. Added test case.

File: tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo [hunk 4]
```diff
@@ -0,0 +1,31 @@
+func test_happy_path() {
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo:1]: [add test](https://github.com/software-mansion/protostar/pull/1384/commits/f28b3f253127ff85a44037432735fc7f0a50e716)

File: protostar/starknet/cheaters/contracts.py (hunk 1 above)
Comment (kasperski95) [reply protostar/starknet/cheaters/contracts.py]: Maybe in a different PR.

File: protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py (hunk 2 above)
Comment (kasperski95) [reply protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py]: I am not sure if that's necessary. I think we should return validation errors instead of raising them.

cc @MaksymilianDemitraszek

File: tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo:1]: This test does not do above. It tests a handler and setting the state inside it with cheatable values (which is of course a valuable test), but doesn't test the logic when l1 handler calls another contract. If it's to be added in the future, please add an issue.

**Review by Arcticae — APPROVED** [review]

File: tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/pure_cairo_vm/cheatcodes/send_message_to_l2/send_message_to_l2_cairo_cheatcode_test.cairo:1]: [add extra test case](https://github.com/software-mansion/protostar/pull/1384/commits/826b3c910144a84ea8a6eb31ec6bc18c3e3d345b)

File: protostar/starknet/cheaters/contracts.py (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply protostar/starknet/cheaters/contracts.py]: Maybe name it L1Cheater?
I agree that Contracts cheater is not a good location

Comment (MaksymilianDemitraszek) [reply protostar/starknet/cheaters/contracts.py]: Can be done in a separate PR, but please create a task for it

Comment (kasperski95) [reply protostar/starknet/cheaters/contracts.py]: I'd have to move the data transformation code out of this class, and that's a subject for a separate PR.

Comment (MaksymilianDemitraszek) [reply protostar/starknet/cheaters/contracts.py]: @karol-bisztyga already did it in #1363 , you can repoint the PR into it

File: protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py]: I opened thread on slack, we should catch those in `Cheatcode`

File: protostar/cheatable_starknet/cheatcodes/send_message_to_l2_cairo_cheatcode.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        ) -> None:
+            asyncio.run(
+                self.cheaters.contracts.send_message_to_l2(
+                    cheaters=self.cheaters,
+                    from_l1_address=Address.from_user_input(from_address),
+                    to_l2_address=Address.from_user_input(to_address),
+                    selector=Selector(function_name),
+                    payload=payload,
+                )
+            )
```
Comment (MaksymilianDemitraszek) [inline protostar/cheatable_starknet/cheatcodes/send_message_to_l2_cairo_cheatcode.py:30]: We should catch data transformer excpetions here, and wrap them in reported exception

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py (hunk 2 above)
Comment (kasperski95) [reply protostar/testing/cairo_cheatcodes/send_message_to_l2_cairo_cheatcode.py]: Blocked by https://github.com/software-mansion/protostar/pull/1404

File: protostar/starknet/cheaters/contracts.py (hunk 1 above)
Comment (kasperski95) [reply protostar/starknet/cheaters/contracts.py]: Removed support for data transformation. I don't know if there will be a resonable way to use it in Cairo 1. If it's needed, it can be added in a different PR because this one hangs already too long.

File: protostar/cheatable_starknet/cheatcodes/send_message_to_l2_cairo_cheatcode.py (hunk 5 above)
Comment (kasperski95) [reply protostar/cheatable_starknet/cheatcodes/send_message_to_l2_cairo_cheatcode.py]: Removed data transformer support from this PR.

File: protostar/cheatable_starknet/cheaters/contracts.py [hunk 6]
```diff
@@ -293,6 +287,54 @@ async def invoke(
...
+            call_type=CallType.DELEGATE,
+            class_hash=await self.cheatable_state.get_class_hash_at(int(to_l2_address)),
+        )
+        with self.cheatable_state.copy_and_apply() as state_copy:
+            await entry_point.execute_for_testing(
+                state=state_copy,
+                general_config=StarknetGeneralConfig(),
+            )
+
+    async def _transform_payload_from_l1_to_cairo_data_by_addr(
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/cheatable_starknet/cheaters/contracts.py]: Removed data transformer support from this PR.

**Comment by kasperski95** [comment]
@MaksymilianDemitraszek ping

**Review by MaksymilianDemitraszek — APPROVED** [review]
